src/puppetlabs/puppetdb/queue.clj

```diff
@@ -190,47 +199,65 @@
    :certname s/Str
    :producer-ts (s/maybe pls/Timestamp)
    :callback (s/=> s/Any s/Any)
-   :command-stream java.io.InputStream})
+   :command-stream java.io.InputStream
+   :compression compression-file-extension-schema})
 
 (s/defn create-command-req :- command-req-schema
   "Validating constructor function for command requests"
   [command :- s/Str
    version :- s/Int
    certname :- s/Str
    producer-ts :- (s/maybe s/Str)
+   compression :- compression-file-extension-schema
    callback :- (s/=> s/Any s/Any)
    command-stream :- java.io.InputStream]
   {:command command
    :version version
    :certname certname
    :producer-ts (when producer-ts
                   (pdbtime/from-string producer-ts))
+   :compression compression
    :callback callback
    :command-stream command-stream})
 
-(defrecord CommandRef [id command version certname received producer-ts callback delete?])
+(defrecord CommandRef [id command version certname received producer-ts callback delete? compression])
 
 (defn cmdref->entry [{:keys [id received] :as cmdref}]
   (stock/entry id (serialize-metadata received cmdref)))
 
 (defn entry->cmdref [entry]
-  (let [{:keys [received command version certname producer-ts]} (-> entry
-                                                                    stock/entry-meta
-                                                                    parse-metadata)]
+  (let [{:keys [received command version
+                certname producer-ts compression]} (-> entry
+                                                       stock/entry-meta
+                                                       parse-metadata)]
     (map->CommandRef {:id (stock/entry-id entry)
                       :command command
                       :version version
                       :certname certname
                       :received received
                       :producer-ts producer-ts
-                      :callback identity})))
+                      :callback identity
+                      :compression compression})))
+
+(defn wrap-decompression-stream
+  [file-extension command-stream]
+  (case file-extension
+    nil command-stream
+    "" command-stream
+    "gz" (GzipCompressorInputStream. command-stream)
+    (throw+ {:kind ::parse-error} nil
+            (trs "Unsupported compression file extension for command: {0}"
+                 file-extension))))
 
 (defn cmdref->cmd [q cmdref]
-  (let [entry (cmdref->entry cmdref)]
+  (let [compression (:compression cmdref)
+        entry (cmdref->entry cmdref)]
     (with-open [command-stream (stock/stream q entry)]
       (assoc cmdref
-             :payload (stream->json command-stream)
-             :entry entry))))
+        :payload (stream->json (wrap-decompression-stream
```

Comment:
@senior That's great to know - thanks for doing that testing.  Happy to add a comment here.

TBH, I'd gone with just wrapping the `GzipCompressorInputStream` around the original input stream - without an extra buffered input stream in between - because I'd noticed that the `GzipCompressorInputStream` already managed its own internal buffer, which it defaults to 8192 bytes.  Double buffering the data before decompressing it intuitively didn't seem like it would bode well for performance.

I think I could have gone with the JDK's `GZIPInputStream` instead of the apache commons one here, for which the internal buffer size is configurable.  I naively went with `GzipCompressorInputStream` instead because we were already using it in PuppetDB - when handling archive files.  Where we probably won't have the requirement of decompressing concatenated `.gz` files in the context of handling PuppetDB commands, though, it may be better to go with `GZIPInputStream` later and do some more targeted buffer size tweaking.  Maybe that would be worth a follow-on effort if it seems worthwhile?